### PR TITLE
Add spatial audio attach to camera, mesh or transform node feature

### PR DESCRIPTION
**Pull request by docEdub**
This change implements the "attach" feature for the spatial audio listener and sound sources.

### CODE REVIEW

**Comment by MrGallow88** [comment]
Thank you.


Sent from Yahoo Mail for iPhone


On Saturday, February 8, 2025, 1:16 PM, Copilot ***@***.***> wrote:




@Copilot commented on this pull request.

Copilot reviewed 5 out of 13 changed files in this pull request and generated 3 comments.
 Files not reviewed (8)    
   - packages/dev/core/src/AudioV2/abstractAudio/subProperties/abstractSpatialAudioListener.ts: Evaluated as low risk
   - packages/dev/core/src/AudioV2/webAudio/webAudioEngine.ts: Evaluated as low risk
   - packages/dev/core/src/AudioV2/webAudio/subNodes/webAudioBusAndSoundSubGraph.ts: Evaluated as low risk
   - packages/dev/core/src/AudioV2/abstractAudio/subProperties/spatialAudioListener.ts: Evaluated as low risk
   - packages/dev/core/src/AudioV2/abstractAudio/subProperties/spatialAudio.ts: Evaluated as low risk
   - packages/dev/core/src/AudioV2/abstractAudio/subNodes/abstractAudioSubGraph.ts: Evaluated as low risk
   - packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/abstractSpatialAudioAttacher.ts: Evaluated as low risk
   - packages/dev/core/src/AudioV2/abstractAudio/subProperties/abstractSpatialAudio.ts: Evaluated as low risk
 Comments suppressed due to low confidence (1) 
packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:9
   
   - [nitpick] The constants in _SpatialAudioAttachedEntity should follow a consistent naming convention. Consider using uppercase for the constants.
export const _SpatialAudioAttachedEntity = {

In packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:
> +                return this._attachedEntity
+                    ? _CreateSpatialAudioCameraAttacherAsync(this._attachedEntity as Camera, this._spatialAudioNode, this._attachmentType, this._minUpdateTime)
+                    : null;
+            case _SpatialAudioAttachedEntity.MESH:
+                return this._attachedEntity
+                    ? _CreateSpatialAudioMeshAttacherAsync(this._attachedEntity as AbstractMesh, this._spatialAudioNode, this._attachmentType, this._minUpdateTime)
+                    : null;
+            case _SpatialAudioAttachedEntity.TRANSFORM_NODE:
+                return this._attachedEntity
+                    ? _CreateSpatialAudioTransformNodeAttacherAsync(this._attachedEntity as TransformNode, this._spatialAudioNode, this._attachmentType, this._minUpdateTime)
+                    : null;
+        }
+        return null;
+    }
+
+    private async _resetAttachedEntity(entity: Nullable<AbstractMesh | Camera | TransformNode>, attacherClassName: string): Promise<void> {

The method _resetAttachedEntity does not handle potential errors from asynchronous operations. Add error handling to ensure robustness.

In packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioCameraAttacher.ts:
> +}
+
+/** @internal */
+export class _SpatialAudioCameraAttacher extends _AbstractSpatialAudioAttacher {
+    protected _camera: Nullable<Camera> = null;
+
+    /** @internal */
+    public constructor(camera: Camera, spatialAudioNode: ISpatialAudioNode, attachmentType: SpatialAudioAttachmentType, minUpdateTime: number) {
+        super(spatialAudioNode, attachmentType, minUpdateTime);
+
+        this.camera = camera;
+    }
+
+    /** @internal */
+    public set camera(camera: Nullable<Camera>) {
+        if (this._camera === this.camera) {

The comparison should be if (this._camera === camera) {.
⬇️ Suggested change-        if (this._camera === this.camera) {
+        if (this._camera === camera) {

In packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioCameraAttacher.ts:
> +    public override dispose(): void {
+        super.dispose();
+        this._clearCamera();
+    }
+
+    /** @internal */
+    public getClassName(): string {
+        return "_SpatialAudioCameraAttacher";
+    }
+
+    /** @internal */
+    public override update(force = false): void {
+        super.update();
+
+        if (force) {
+            this._camera?.computeWorldMatrix();

The computeWorldMatrix method should be called with the force parameter: this._camera?.computeWorldMatrix(force);.
⬇️ Suggested change-            this._camera?.computeWorldMatrix();
+            this._camera?.computeWorldMatrix(force);

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        this._attachmentType = attachmentType;
+        this._spatialAudioNode = spatialAudioNode;
+    }
+
+    /**
+     * The scene that the audio listener or source is attached to, or null if the audio listener or source is not
+     * attached to a scene.
+     */
+    public get attachedCamera(): Nullable<Camera> {
+        return this._attacher?.getClassName() === _SpatialAudioAttachedEntity.CAMERA ? (this._attachedEntity as Camera) : null;
```
Comment (deltakosh) [inline packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:48]: how does it work for ArcRotateCamera for instance?

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+        this._isReadyPromise = this._resetAttachedEntity(value, _SpatialAudioAttachedEntity.CAMERA);
+    }
+
+    /**
+     * The mesh that the audio listener or source is attached to, or null if the audio listener or source is not
+     * attached to a mesh.
+     */
+    public get attachedMesh(): Nullable<AbstractMesh> {
+        return this._attacher?.getClassName() === _SpatialAudioAttachedEntity.MESH ? (this._attachedEntity as AbstractMesh) : null;
```
Comment (deltakosh) [inline packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:64]: same for instancemesh, etc...

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+        this._isReadyPromise = this._resetAttachedEntity(value, _SpatialAudioAttachedEntity.MESH);
+    }
+
+    /**
+     * The transform node that the audio listener or source is attached to, or null if the audio listener or source is
+     * not attached to a transform node.
+     */
+    public get attachedTransformNode(): Nullable<TransformNode> {
+        return this._attacher?.getClassName() === _SpatialAudioAttachedEntity.TRANSFORM_NODE ? (this._attachedEntity as TransformNode) : null;
```
Comment (deltakosh) [inline packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:80]: as mesh inherits from transformNode, do we need a distinction? can we just support transformNode?

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioCameraAttacher.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        return this._camera?.globalPosition ?? Vector3.ZeroReadOnly;
+    }
+
+    protected get _attachedRotationQuaternion(): Quaternion {
+        if (!this._camera) {
+            TempQuaternion.copyFromFloats(0, 0, 0, 1);
+            return TempQuaternion;
+        }
+
+        this._camera.getWorldMatrix().decompose(undefined, TempQuaternion, undefined);
```
Comment (deltakosh) [inline packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioCameraAttacher.ts:58]: won't that be a problem to share the same quaternion among several entities? what if I have 2 cameras or 2 meshes with sound? they will overwrite their quaternion value

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts (hunk 3 above)
Comment (docEdub) [reply packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:80]: The old audio engine centers the spatial audio in the mesh's bounding box, which is what `attachedMesh` does in the new audio engine. It will be useful when the old audio engine is reimplemented with the new audio engine.

Are you thinking it's not useful for users? Maybe it should be kept but marked as internal?

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioCameraAttacher.ts (hunk 4 above)
Comment (docEdub) [reply packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioCameraAttacher.ts:58]: I don't think so. A copy is made when setting the result on the spatial audio node:

https://github.com/BabylonJS/Babylon.js/blob/38b39538e67a49396bac8b0a6f8e49782b70b325/packages/dev/core/src/AudioV2/webAudio/subNodes/spatialWebAudioSubNode.ts#L150-L151

We could make another copy to be safe, but since this is only used internally and it's copied in the end I think it's ok to just return `TempQuaternion` here.

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts (hunk 2 above)
Comment (docEdub) [reply packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:64]: The current audio engine uses the mesh's bounding box, so we've got that same functionality covered in the new audio engine with the `attachedMesh` property.

Comment (deltakosh) [reply packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:64]: yes but you are testing agains "mesh" on class name and I think you may want to use instanceof in that case

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts (hunk 1 above)
Comment (docEdub) [reply packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:48]: Oh, I see what you mean. Thanks for clarifying. The class name being referenced here is for the `_attacher` helper, not the camera.

File: packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts (hunk 2 above)
Comment (docEdub) [reply packages/dev/core/src/AudioV2/abstractAudio/spatialAttachers/spatialAudioAttacher.ts:64]: Same as above comment. It's the `_attacher` helper's class name being compared, so it should work fine with any `AbstractMesh` subclass.

**Comment by RaananW** [comment]
> Just thinking out loud, so please completely dismiss if it does not make sense.
> 
> I wonder whether setOptions should be async ? it looks like it does way more than setting options. Also it seems like we are adding quite a few convenience shortcuts but I am wondering if they are need or if the _attacher should be exposed ?
> 
> As I said I know nothing about audio so do not mind me too much.

I think that's a valid point. What is the reasoning behind making setOptions async?

File: packages/dev/core/src/AudioV2/abstractAudio/subProperties/spatialAudioListener.ts [hunk 5]
```diff
@@ -1,20 +1,134 @@
...
+        this._attacher.detach();
+    }
+
+    /** @internal */
+    public dispose(): void {
+        this._attacher.dispose();
+    }
+
+    /** @internal */
+    public async setOptions(options: Partial<ISpatialAudioListenerOptions>): Promise<void> {
```
Comment (RaananW) [inline packages/dev/core/src/AudioV2/abstractAudio/subProperties/spatialAudioListener.ts:95]: A comment made by Seb - why is this async now? there is nothing async in the function body (unless I missed something?). Or is it due to inheritance from a different branch?

The same goes to the other setOptions function changed in this PR.

**Review by RaananW — APPROVED** [review]
Approving, pending an answer and resolution on the setOptions function.

Reviewers, please note - this is not merging into master. This branch is merging to the audio-engine branch! 

Please let @docEdub make the decisions regarding when to merge this.

**Comment by RaananW** [comment]
@docEdub - at our last conversation you said you want to change the viewMatrix to the camera's world matrix instead, but I just noticed this was not pushed yet. I assume you want to first make the changes before this is approved?

**Review by sebavan — CHANGES_REQUESTED** [review]
Flagging as change requested as it seems this one could go without all comments resolved

**Review by sebavan — COMMENTED** [review]
Nah... I was just being dumb... and did not notice the push to a branch...

**Comment by docEdub** [comment]
Thanks for the reviews. I will make the requested changes and create a new PR against the `master` branch now that the main audio engine V2 PR is merged.

File: packages/dev/core/src/AudioV2/abstractAudio/subProperties/spatialAudioListener.ts (hunk 5 above)
Comment (docEdub) [reply packages/dev/core/src/AudioV2/abstractAudio/subProperties/spatialAudioListener.ts:95]: They are both async because they await `this._attacher.isReadyPromise;`.

**Comment by MrGallow88** [comment]
Thank you.


Sent from Yahoo Mail for iPhone


On Tuesday, March 4, 2025, 9:00 PM, Drew Fillebrown ***@***.***> wrote:




@docEdub commented on this pull request.

In packages/dev/core/src/AudioV2/abstractAudio/subProperties/spatialAudioListener.ts:
> +    }
+
+    /**
+     * Detaches the audio source from the currently attached camera, mesh or transform node.
+     */
+    public detach(): void {
+        this._attacher.detach();
+    }
+
+    /** @internal */
+    public dispose(): void {
+        this._attacher.dispose();
+    }
+
+    /** @internal */
+    public async setOptions(options: Partial<ISpatialAudioListenerOptions>): Promise<void> {

They are both async because they await this._attacher.isReadyPromise;.
